### ISSUE TITLE
verify if the expiration date of an SMS token is handled correctly

### ISSUE THREAD

**Issue by cornelinux** [privacyidea/privacyidea#496]
A question about the experiation date on the mailing list:
https://groups.google.com/forum/#!topic/privacyidea/4a4KN3UDoaE

<pre>
I've requested token by SMS and received it.

This is the "List Challenges":

Wed, 24 Aug 2016 09:33:28 GMT     PISM0000CC7B     00155976067561524466     Wed, 24 Aug 2016 09:38:28 GMT     0

Now... I'm waiting more than 5 minutes.....
And then I try to validate token.. (09:43)

Result.. 

"matching 1 tokens", "type": "sms", "serial": "PISM0000CC7B", "otplen": 6}
"result": {"status": true, "value": true}, "time": 1472024584.399257, "id": 1}

Mmmm....

This is the log

11572 - 8/24/16 09:43:04 - GET /validate/check - 1 - PISM0000CC7B - ... matching 1 tokens - OK

But the Challenges report expiration : 09:38:28
</pre>

We need to check this. It seems as if the tokenclass is_valid() method is not touched.

**Comment by cornelinux** [comment]
Work on a library level. After fixing the **test** *test_lib_tokens_sms* is_valid() is perfectly called and it behaves perfectly.

**Comment by cornelinux** [comment]
The challenge is only checked, when authenticating with the transaction ID.
When authenticating

 1. step: PIN
  -> OTP send

 2. step: PIN+OTP

the expiration will not be checked, since: The sent OTP value is an HOTP value. And if the PIN is sent togeather with the OTP, then this is a successful single-shot-authentication and Challenge/Response will not be checked.

**Comment by cornelinux** [comment]
At the moment I see no need to change this behaviour

**Comment by klinux** [comment]
Hi Cornelinux,

I have a doubt about otp expiry, if I check otp using PIN+OTP, the expiration data of OTP code is not checked? What is the life time of that OTP code? Is there anyway to validate expiry time of that type authentication?
Thank you.

**Comment by cornelinux** [comment]
> I have a doubt about otp expiry, if I check otp using PIN+OTP, the expiration date of OTP code is not  checked? 

The OTP code has not "expiration date". The expiration date is in the challenge, i.e. the user presented his OTP PIN. And this is why he is allowed to login for an ``expiration time`` with only the OTP value.

> What is the life time of that OTP code?

eternity

> Is there anyway to validate expiry time of that type authentication?

I do not understand that question.

By the way: I very much recommend that you do not use SMS anyway. It is the unsafest way to authenticate which you can imagine. Ask NIST.
Closely followed by the Google Authenticator. ;-)

**Comment by klinux** [comment]
Thank you for the answer,

Let me explain my problem here.

I'm using only the serial and pin to create a otp code, the user receive the sms with otp,everything works fine, but, we receive a "valid_until": "2017-07-21 06:23:06.709327" for the challenge, the challenge has only 5 minutes to be checked, but after 10 minutes I had success with the otp code.

I'm using the follow payload to validate: {"serial": "PISM0000F1C3", "pass": "PIN+OTP"}, the period validate of the challenge wasn't respected, I'm doing something wrong?

Thank you.

**Comment by cornelinux** [comment]
Yes.
The way challenge/response in privacyIDEA is ment to work is this way:
http://privacyidea.readthedocs.io/en/latest/modules/lib/tokentypes/sms.html?highlight=challenge%20response

Your second request is not supposed to contain the PIN, but the transaction id.

**Comment by klinux** [comment]
Cornelinux,

I have a question about the challenge time validity, how can I change this time? I changed the DefaultChallengeValidityTime, but the time of the challenge is always created with 5 minutes. How can I change that to 10 minutes?

Thank you.

**Comment by cornelinux** [comment]
> I have a question about the challenge time validity, how can I change this time? I changed the DefaultChallengeValidityTime, but the time of the challenge is always created with 5 minutes. How can I change that to 10 minutes?

Each token type has its own ``create_challenge`` function. Which token type are you talking about?

> In the case, is a sms serial, when I use email serial the valid_until is not returned, only in sms serial that time is returned.

I do not understand this. Can be please explain in more detail where it is returned?

**Comment by klinux** [comment]
I'm create a token via api rest /token/init with payload like that:
{
    "timeStep":30,
    "otplen":6,
    "genkey":true,
    "type":"sms",
    "hashlib":"sha1",
    "radius.system_settings":true,
    "phone":"1195202XXXX",
    "serial":"c2ea5672-e225-11e4-8a00-1681e6b8833",
    "totp.hashlib":"sha1",
    "hotp.hashlib":"sha1",
    "pin":"123456"
}

And send the otp via sms via api rest /validate/check with payload like that:
{
	"serial":"c2ea5672-e225-11e4-8a00-1681e6b8833",
	"pass":"123456"
}

And I get the response from the server:
{
    "jsonrpc": "2.0",
    "signature": "1475821792...",
    "detail": {
        "multi_challenge": [
            {
                "attributes": {
                    "valid_until": "**2018-02-16 15:30:03.803872**",
                    "state": null
                },
                "serial": "c2ea5672-e225-11e4-8a00-1681e6b8833",
                "transaction_id": "10439365256273624132"
            }
        ],
        "threadid": 140348240893696,
        "attributes": {
            "valid_until": "**2018-02-16 15:30:03.803872**",
            "state": null
        },
        "message": "Enter the OTP from the SMS:",
        "serial": "c2ea5672-e225-11e4-8a00-1681e6b8833",
        "transaction_id": "10439365256273624132"
    },
    "versionnumber": "2.21.4",
    "version": "privacyIDEA 2.21.4",
    "result": {
        "status": true,
        "value": false
    },
    "time": 1518801903.924229,
    "id": 1
}

The valid_until is always set with 5 minutes.

And email token do not set the valid_until in the response. But email token I can change the time expire in the web ui, after 10 minutes, the otp validation failed, like expected.

Thank you

**Comment by klinux** [comment]
I found here where is my problem.

The method **get_from_config("sms.providerTimeout", 5 * 60)** wait the sms.providerTimeout from database, but this config won't set in the database, so, the validty time is set with 5 * 60, 5 minutes.

In the sms token configuration page, the sms.providerTimeout (Timeout label) input is disabled below the **SMS Token settings**.

I set the sms.providerTimeout in the database, now valid_until is OK.

Why this input is disabled?

Thank you
